File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlPivotClause.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';

export class SqlPivotClause implements ISqlExpression {
    SqlType: SqlType = SqlType.PivotClause;
    Span: TextSpan = new TextSpan();
    NewColumn!: ISqlExpression;
    ForSource!: ISqlExpression;
    InColumns: ISqlExpression[] = [];
    AliasName: string = '';

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_PivotClause(this);
    }

    ToSql(): string {
        let sql = 'PIVOT (\n';
        sql += this.NewColumn.ToSql() + '\n';
        sql += 'FOR ' + this.ForSource.ToSql() + '\n';
        sql += 'IN (' + this.InColumns.map(x => x.ToSql()).join(', ') + ') ';
        sql += `AS ${this.AliasName}`;
        sql += ')';
        return sql;
    }
} 

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlPivotClause.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlPivotClause : ISqlExpression
{
    public SqlType SqlType => SqlType.PivotClause;
    public TextSpan Span { get; set; } = new();
    public void Accept(SqlVisitor visitor)
    {
        visitor.Visit_PivotClause(this);
    }

    public required ISqlExpression NewColumn { get; set; }
    public required ISqlExpression ForSource { get; set; }
    public List<ISqlExpression> InColumns { get; set; } = [];
    public string AliasName { get; set; } = string.Empty; 

    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write("PIVOT (");
        sql.WriteLine();
        sql.Indent++;
        sql.Write(NewColumn.ToSql());
        sql.WriteLine();
        sql.Write("FOR ");
        sql.Write(ForSource.ToSql());
        sql.WriteLine();
        sql.Write("IN (");
        sql.WriteLine();
        sql.Indent++;
        foreach (var item in InColumns.Select((column, index) => new { column, index }))
        {
            sql.Write(item.column.ToSql());
            if (item.index < InColumns.Count - 1)
            {
                sql.Write(",");
            }
            sql.WriteLine();
        }
        sql.Indent--;
        sql.Write(") ");
        sql.WriteLine($"AS {AliasName}");
        sql.Indent--;
        return sql.ToString();
    }
}